(DOCSP-25080) Add sync section to JS quickstarts (#2365)
mongodb/docs-realm: source/sdk/react-native/quick-start.txt

==============================
Quick Start - React Native SDK
==============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This page contains information to quickly get Realm Database
integrated into your app.

If you haven't already, :ref:`install the Realm React Native SDK <react-native-install>`.

.. _react-native-quickstart-local-import-realm:

Import Realm
------------
At the top of your source files where you want to use Realm, add
the following line to import the SDK.

.. literalinclude:: /examples/generated/node/quickstart.snippet.import-realm.js
  :language: javascript

.. _react-native-quickstart-local-define-an-object-model:

Define Your Object Model
------------------------
Your application's :ref:`object model <react-native-object-schemas>` defines the data
that you can store within Realm Database. 

To define a Realm object type, create a schema object that specifies the type's
``name`` and ``properties``. The type name must be unique among object types in
a realm.

The following code shows how to define an object model for a ``Task`` object. In this example:
  

> - The ``primaryKey`` is the ``_id`` of type ``int`` (Another common type used for primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`).

Reviewer: ```suggestion
- The ``primaryKey`` is the ``_id`` of type ``int``. (Another common type used for primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`.)
```

your App Services App using your App ID. You can :ref:`find your App ID <find-your-app-id>`
in the App Services UI.

.. literalinclude:: /examples/generated/node/quickstart.snippet.initialize.js
  :language: javascript

Authenticate a User
~~~~~~~~~~~~~~~~~~~

To authenticate and log in a user, call :js-sdk:`App.logIn()
<Realm.App.html#logIn>`. When :ref:`anonymous authentication
<anonymous-authentication>` is enabled, users can immediately log into your app
without providing any identifying information:

.. literalinclude:: /examples/generated/node/quickstart.snippet.anonymous-login.js
  :language: javascript

Define an Object Model
~~~~~~~~~~~~~~~~~~~~~~

Object models for synced realms work the same way as :ref:`local-only Realms 
<react-native-quickstart-local-define-an-object-model>`. Define your object model 
just as you would for a local-only Realm.

.. literalinclude:: /examples/generated/node/quickstart.snippet.define-an-object-model.js
  :language: javascript

Open a Synced Realm
~~~~~~~~~~~~~~~~~~~

After you have initialized your App, authenticated a user, and
defined your object model, you can create a :js-sdk:`SyncConfiguration

> <Realm.App.Sync.html#~SyncConfiguration>`.

Reviewer: need to be clearer that `SyncConfiguration` is just part of the larger `Configuration` object. see the text at the beginning of this section - https://www.mongodb.com/docs/realm/sdk/node/examples/open-and-close-a-realm/#open-a-flexible-synced-realm

   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This page contains information to quickly get Realm Database
integrated into your app.

If you haven't already, :ref:`install the Realm React Native SDK <react-native-install>`.

.. _react-native-quickstart-local-import-realm:

Import Realm
------------
At the top of your source files where you want to use Realm, add
the following line to import the SDK.

.. literalinclude:: /examples/generated/node/quickstart.snippet.import-realm.js
  :language: javascript

.. _react-native-quickstart-local-define-an-object-model:

Define Your Object Model
------------------------
Your application's :ref:`object model <react-native-object-schemas>` defines the data
that you can store within Realm Database. 

To define a Realm object type, create a schema object that specifies the type's
``name`` and ``properties``. The type name must be unique among object types in
a realm.

The following code shows how to define an object model for a ``Task`` object. In this example:
  
- The ``primaryKey`` is the ``_id`` of type ``int``. (Another common type used for 
  primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`.)

> - The ``name`` field is required. The

Reviewer: ```suggestion
- The ``name`` field is required.
```

> - The ``status`` is optional, denoted by the question mark immediately after the data type.

Reviewer: ```suggestion
- The ``status`` and ``onwer_id`` fields are optional, denoted by the question mark immediately after the data type.
```

<Realm.App.html#logIn>`. When :ref:`anonymous authentication
<anonymous-authentication>` is enabled, users can immediately log into your app
without providing any identifying information:

.. literalinclude:: /examples/generated/node/quickstart.snippet.anonymous-login.js
  :language: javascript

Define an Object Model
~~~~~~~~~~~~~~~~~~~~~~

Object models for synced realms work the same way as :ref:`local-only Realms 
<react-native-quickstart-local-define-an-object-model>`. Define your object model 
just as you would for a local-only Realm.

.. literalinclude:: /examples/generated/node/quickstart.snippet.define-an-object-model.js
  :language: javascript

Open a Synced Realm
~~~~~~~~~~~~~~~~~~~

After you have initialized your App, authenticated a user, and
defined your object model, you can create a :js-sdk:`SyncConfiguration
<Realm.App.Sync.html#~SyncConfiguration>`.

To open a Flexible Sync realm, call :js-sdk:`Realm.open() <Realm.html#.open>`. 
Pass in a :js-sdk:`Configuration <Realm.html#~Configuration>`
object, which must include the ``sync`` property defining a 
:js-sdk:`SyncConfiguration <Realm.App.Sync.html#~SyncConfiguration>` object. 

> In the SyncConfiguration, you must include include a ``user`` and ``flexible:true``.

Reviewer: ```suggestion
In the SyncConfiguration, you must include include a ``user`` and ``flexible: true``.
```

Additionally, you need at least one subscription before you can read from or write to the
realm. Use ``Configuration.sync.initialSubscriptions`` to define the initial 
subscription set when the Realm file is first opened.

.. literalinclude:: /examples/generated/node/quickstart.snippet.open-realm-with-subscriptions.js

>   :language: javascript

Author: Sorry for the misunderstanding! I've fixed this in my latest commit.

.. _react-native-client-quick-start-local:
.. _react-native-client-quick-start:

==============================
Quick Start - React Native SDK
==============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This page contains information to quickly get Realm Database
integrated into your app.

.. _react-native-quickstart-local-import-realm:

> 

Reviewer: another perhaps out of scope comment, but a good quick win. there should be some info abt installing the sdk on this page. could be as simple as a bit of text and link to install page. 

see flutter sdk quick start for example - https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/flutter/quick-start/

Author: Good point. I'll include that in my next commit.

Flexible Sync realm to sync data between devices.

Prerequisites
~~~~~~~~~~~~~

- :ref:`An App Services App <create-a-realm-app>`
- :ref:`Anonymous authentication <anonymous-authentication>` enabled in the App Services UI
- :ref:`Flexible Sync <enable-flexible-sync>` enabled with :ref:`Development Mode <development-mode>`
  on and an ``owner_id`` field in the :guilabel:`Queryable Fields` section

Initialize the App
~~~~~~~~~~~~~~~~~~

To use App Services features, such as authentication and sync, you must first access
your App Services App using your App ID. You can :ref:`find your App ID in
the App Services UI <find-your-app-id>`.

.. code-block:: javascript

   import Realm from "realm";

   const appId = "<yourAppId>"; // Set App ID here.
   const appConfig = {
     id: appId,
     timeout: 10000
   };
   const app = new Realm.App(appConfig);

Authenticate a User
~~~~~~~~~~~~~~~~~~~

To authenticate and log in a user, call :js-sdk:`App.logIn()
<Realm.App.html#logIn>`. When :ref:`anonymous authentication
<anonymous-authentication>` is enabled, users can immediately log into your app
without providing any identifying information:

.. code-block:: javascript

>    :emphasize-lines: 8,9

Reviewer: ```suggestion
```

this doesn't make sense anymore

   import Realm from "realm";

   const appId = "<yourAppId>"; // Set App ID here.
   const appConfig = {
     id: appId,
     timeout: 10000
   }; 

   async function anonymousLogin() {

>      let user;

Reviewer: there's no reason why this `user` should be declared outside the try block b/c it's only used within it.

```suggestion
```

     try {

>        const app = new Realm.App(appConfig); // pass in the appConfig variable that you created earlier

Reviewer: ```suggestion
       // pass in the appConfig variable that you created earlier
       const app = new Realm.App(appConfig);
```

nit: line was long and getting cut off. now easier to read

       const credentials = Realm.Credentials.anonymous(); // create an anonymous credential

>        user = await app.logIn(credentials);

Reviewer: ```suggestion
       const user = await app.logIn(credentials);
```

       return user;

     } catch (error) {
         throw `Error logging in anonymously: ${JSON.stringify(error,null,2)}`;
     }
   }

Open a Synced Realm
~~~~~~~~~~~~~~~~~~~

After you have initialized your App, authenticated a user, and
defined your object model, you can create a :js-sdk:`SyncConfiguration

> <Realm.App.Sync.html#~SyncConfiguration>`:

Reviewer: ```suggestion
<Realm.App.Sync.html#~SyncConfiguration>`.
```

If you have opened a local realm following the :ref:`Open a Realm
<react-native-quickstart-local-open-a-realm>` section above, replace the
``RealmConfiguration`` with the ``SyncConfiguration`` described below.

Pass the authenticated user and the ``Item`` class to the ``SyncConfiguration``
function to create a :ref:`Flexible Sync Configuration
<react-native-flexible-sync>`.

You need at least one subscription before you can read from or write to the
realm. Use ``Configuration.sync.initialSubscriptions`` to define the initial subscription set when
the Realm file is first opened. Pass the query you wish to subscribe to and a
name for the subscription to the :js-sdk:`add()
<Realm.App.Sync.MutableSubscriptionSet.html#add>` method on the
``MutableSubscriptionSet``.

.. code-block:: javascript
   :emphasize-lines: 8,9

>    const config = {

Reviewer: this config doesn't have the schema. 

i think you should just bluehawk the code examples here to reduce risk of error like this

.. _react-native-client-quick-start-local:
.. _react-native-client-quick-start:

==============================
Quick Start - React Native SDK
==============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This page contains information to quickly get Realm Database
integrated into your app.

.. _react-native-quickstart-local-import-realm:

Import Realm
------------
At the top of your source files where you want to use Realm, add
the following line to import the SDK.

.. literalinclude:: /examples/generated/node/quickstart-local.snippet.import-realm.js
  :language: javascript

.. _react-native-quickstart-local-define-an-object-model:

Define Your Object Model
------------------------
Your application's :ref:`object model <react-native-object-schemas>` defines the data
that you can store within Realm Database. 

To define a Realm object type, create a schema object that specifies the type's
``name`` and ``properties``. The type name must be unique among object types in
a realm.

> .. example::

Reviewer: i recommend removing all `.. example::` directives on this page. that's generally what we've done across the docs. 

also seems like some examples on this page aren't in the directive as well.

   The following code shows how to define an object model for a ``Task`` object. In this example:
     
   - The ``primaryKey`` is the ``_id`` of type ``int`` (Another common type used for primary keys is :js-sdk:`ObjectId <Realm.Object.html#objectId>`).
   - The ``name`` field is required. The
   - The ``status`` is optional, denoted by the question mark immediately after the data type.
    
   .. literalinclude:: /examples/generated/node/quickstart-local.snippet.define-an-object-model.js
     :language: javascript

.. _react-native-quickstart-local-open-a-realm:

Open a Realm
------------
To open a realm, pass a :js-sdk:`Realm.Configuration
<Realm.html#~Configuration>` object to either :js-sdk:`Realm.open()
<Realm.html#.open>` or :js-sdk:`new Realm() <Realm.html>`.

.. literalinclude:: /examples/generated/node/quickstart-local.snippet.open-a-realm.js
  :language: javascript

.. _react-native-quickstart-local-create-realm-objects:

Create Realm Objects
--------------------

Once you have opened a realm, you can create :ref:`objects
<react-native-object-types>` in it. All writes must occur within a :ref:`Write
transaction <react-native-write-transactions>`.

.. example::

>    In the following example, an application developer creates two tasks. One task
>    with a name of "go grocery shopping" and another with the name of "go
>    exercise". Both tasks have a status of "Open".

Reviewer: ```suggestion
```

don't think this provides any value to reader

    
   .. literalinclude:: /examples/generated/node/quickstart-local.snippet.create-realm-objects.js
     :language: javascript

.. _react-native-quickstart-local-find-sort-and-filter-objects:

Find, Sort, and Filter Objects

> ------------------------------

Reviewer: ```suggestion
------------------------------

```

   example above. 
    
   .. literalinclude:: /examples/generated/node/quickstart-local.snippet.delete-an-object.js
     :language: javascript

.. _react-native-quickstart-local-watch-a-collection:

Watch a Collection
------------------
You can :ref:`watch a realm, collection, or object for changes
<react-native-change-notifications>` by registering event handlers with the
:js-sdk:`Realm.addListener() <Realm.html#addListener>`
:js-sdk:`Object.addListener() <Realm.Object.html#addListener>`
:js-sdk:`Collection.addListener() <Realm.Collection.html#addListener>`
methods.

.. example::

   In the following example, an application developer watches for changes to the
   ``Task`` collection.
    
   .. literalinclude:: /examples/generated/node/quickstart-local.snippet.watch-a-collection.js
     :language: javascript

.. _react-native-quickstart-local-close-a-realm:

Close a Realm
-------------
Call the :js-sdk:`realm.close() <Realm.html#close>` method when done with a realm instance to avoid memory leaks.

.. literalinclude:: /examples/generated/node/quickstart-local.snippet.close-a-realm.js
  :language: javascript

.. _react-native-quickstart-local-complete:

Complete Example

> ----------------

Reviewer: we've removed the Complete Example sections from the other SDKs. should do so here as well

     "status": "InProgress"
   }
   
   A task was deleted at the 0 index

Add Device Sync (Optional)
--------------------------

This section illustrates how to authenticate with an Anonymous User and open a
Flexible Sync realm to sync data between devices.

Prerequisites
~~~~~~~~~~~~~

- :ref:`An App Services App <create-a-realm-app>`
- :ref:`Anonymous authentication <anonymous-authentication>` enabled in the App Services UI
- :ref:`Flexible Sync <enable-flexible-sync>` enabled with :ref:`Development Mode <development-mode>`
  on and an ``owner_id`` field in the :guilabel:`Queryable Fields` section

Initialize the App
~~~~~~~~~~~~~~~~~~

To use App Services features, such as authentication and sync, you must first access
your App Services App using your App ID. You can :ref:`find your App ID in
the App Services UI <find-your-app-id>`.

.. code-block:: javascript

  import Realm from 'realm';

   // Returns the shared instance of the Realm app.
   export function getRealmApp() {
      const appId = '<enter your App ID here>'; // Set App ID here.
      const appConfig = {
        id: appId,
        timeout: 10000,
      };
     return new Realm.App(appConfig);

>    }

Reviewer: this is a _funky_ way of instantiating the Realm.App instance.

also i think the comment is wrong.

this is creating a _new_ instance of the app every time you call the function. 

i think it'll still work, but it's odd and wastes compute resources reinstantiating the object every time you call the function.

rather you could just do:

```js
const appId = '<enter your App ID here>'; // Set App ID here.
const appConfig = {
  id: appId,
  timeout: 10000
};
const app = new Realm.App(appConfig);
```

Authenticate a User
~~~~~~~~~~~~~~~~~~~

To authenticate and log in a user, call :js-sdk:`App.logIn()
<Realm.App.html#logIn>`. When :ref:`anonymous authentication
<anonymous-authentication>` is enabled, users can immediately log into your app
without providing any identifying information:

.. code-block:: javascript
   :emphasize-lines: 8,9

   import Realm from 'realm';    

   async function anonymousLogin() {
     let user;
     try {

>        const app = new Realm.App(appConfig); // pass in the appConfig variable that you created earlier

Reviewer: this line is also _funky_ b/c it uses the `appConfig` variable which was scoped to the `getRealmApp()` function defined above.

this example should be adjusted based on changes made to above code example

       const credentials = Realm.Credentials.anonymous(); // create an anonymous credential
       user = await app.logIn(credentials);
       return user;

     } catch (error) {
         throw `Error logging in anonymously: ${JSON.stringify(error,null,2)}`;
     }
   }

Open a Synced Realm
~~~~~~~~~~~~~~~~~~~

After you have initialized your App, authenticated a user, and
defined your object model, you can create a :js-sdk:`SyncConfiguration
<Realm.App.Sync.html#~SyncConfiguration>`:

If you have opened a local realm following the :ref:`Open a Realm
<react-native-quickstart-local-open-a-realm>` section above, replace the

> ``RealmConfiguration `` with the ``SyncConfiguration`` described below.

Reviewer: ```suggestion
``RealmConfiguration`` with the ``SyncConfiguration`` described below.
```

fixes rendering issue

Pass the authenticated user and the ``Item`` class to the ``SyncConfiguration``
function to create a :ref:`Flexible Sync Configuration
<react-native-flexible-sync>`.

You need at least one subscription before you can read from or write to the

> realm. Use ``initialSubscriptions`` to define the initial subscription set when

Reviewer: ```suggestion
realm. Use ``Configuration.sync.initialSubscriptions`` to define the initial subscription set when
```

the Realm file is first opened. Pass the query you wish to subscribe to and a
name for the subscription to the :js-sdk:`add()
<Realm.App.Sync.MutableSubscriptionSet.html#add>` method on the
``MutableSubscriptionSet``.

The example below specifies a subscription named "User's Items" with
all ``Item`` objects.

.. code-block:: javascript
   :emphasize-lines: 8,9

   const config = {
     sync: {
       user: app.currentUser,
       flexible: true,
       initialSubscriptions: {
         update: (subs, realm) => {
           subs.add(

>              realm.objects("Item").filtered("name == 'User's Items'")

Reviewer: this query doesn't make particular sense b/c why would multiple `Items` have the name `'User's Items'`?

rather could be smthn like:

```js
realm.objects("Item").filtered("ownerId == $0", app.currentUser.id)
```

           );
         },
       },
     },
   };
   const realm = await Realm.open(config);

> 

Reviewer: i think would be good to follow this example with at least a sentence saying that now can write to realm and data syncs automatically to Atlas.

also data matching the subscription query + backend query filters automatically syncs down to the device.

Reviewer: swift sdk has some good text you can borrow here - https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-25080/sdk/swift/quick-start/#open-a-realm-1

Next: Check out the Template Apps and Tutorial
----------------------------------------------

Check out the :ref:`template app <template-apps>` to experiment with another
quick way to start coding with the Realm React Native SDK. The React Native SDK template,
labeled ``react-native.todo.flex``, is a premade application that integrates
Realm Database and Atlas Device Sync into a customizable android app.

Alternatively, if you are interested in a guided experience, you can read our
:ref:`Android with Kotlin SDK tutorial <react-native-tutorial>` that expands on

> the template app.

Reviewer: merp, react native pls
